lib/internal/streams/web.js

```diff
@@ -0,0 +1,4632 @@
+      if (!writableStreamCloseQueuedOrInFlight(stream) &&
+          stream[kState].backpressure) {
+        writer[kState].ready = createDeferredPromise();
+      } else {
+        writer[kState].ready = {
+          promise: PromiseResolve(),
+          resolve: undefined,
+          reject: undefined,
+        };
+      }
+      setClosedPromiseToNewPromise();
+      break;
+    case 'erroring':
+      writer[kState].ready = {
+        promise: PromiseReject(stream[kState].storedError),
+        resolve: undefined,
+        reject: undefined,
+      };
+      setPromiseHandled(writer[kState].ready.promise);
+      setClosedPromiseToNewPromise();
+      break;
+    case 'closed':
+      writer[kState].ready = {
+        promise: PromiseResolve(),
+        resolve: undefined,
+        reject: undefined,
+      };
+      writer[kState].close = {
+        promise: PromiseResolve(),
+        resolve: undefined,
+        reject: undefined,
+      };
+      break;
+    default:
+      writer[kState].ready = {
+        promise: PromiseReject(stream[kState].storedError),
+        resolve: undefined,
+        reject: undefined,
+      };
+      writer[kState].close = {
+        promise: PromiseReject(stream[kState].storedError),
+        resolve: undefined,
+        reject: undefined,
+      };
+      setPromiseHandled(writer[kState].ready.promise);
+      setPromiseHandled(writer[kState].close.promise);
+  }
+
+  function setClosedPromiseToNewPromise() {
+    writer[kState].close = createDeferredPromise();
+  }
+}
+
+function writableStreamAbort(stream, reason) {
+  const {
+    state,
+  } = stream[kState];
+  if (state === 'closed' || state === 'errored')
+    return PromiseResolve();
+
+  if (stream[kState].pendingAbortRequest.abort.promise !== undefined)
+    return stream[kState].pendingAbortRequest.abort.promise;
+
+  assert(state === 'writable' || state === 'erroring');
+
+  let wasAlreadyErroring = false;
+  if (state === 'erroring') {
+    wasAlreadyErroring = true;
+    reason = undefined;
+  }
+
+  const abort = createDeferredPromise();
+
+  stream[kState].pendingAbortRequest = {
+    abort,
+    reason,
+    wasAlreadyErroring,
+  };
+
+  if (!wasAlreadyErroring)
+    writableStreamStartErroring(stream, reason);
+
+  return abort.promise;
+}
+
+function writableStreamClose(stream) {
+  const {
+    state,
+    writer,
+    backpressure,
+    controller,
+  } = stream[kState];
+  if (state === 'closed' || state === 'errored') {
+    return PromiseReject(
+      new ERR_INVALID_STATE.TypeError('WritableStream is closed'));
+  }
+  assert(state === 'writable' || state === 'erroring');
+  assert(!writableStreamCloseQueuedOrInFlight(stream));
+  stream[kState].closeRequest = createDeferredPromise();
+  const { promise } = stream[kState].closeRequest;
+  if (writer !== undefined && backpressure && state === 'writable')
+    writer[kState].ready.resolve?.();
+  writableStreamDefaultControllerClose(controller);
+  return promise;
+}
```

Comment:
```suggestion
async function writableStreamClose(stream) {
  const {
    state,
    writer,
    backpressure,
    controller,
  } = stream[kState];
  if (state === 'closed' || state === 'errored') {
    throw new ERR_INVALID_STATE.TypeError('WritableStream is closed');
  }
  assert(state === 'writable' || state === 'erroring');
  assert(!writableStreamCloseQueuedOrInFlight(stream));
  stream[kState].closeRequest = createDeferredPromise();
  const { promise } = stream[kState].closeRequest;
  if (writer !== undefined && backpressure && state === 'writable')
    writer[kState].ready.resolve?.();
  writableStreamDefaultControllerClose(controller);
  return promise;
}
```

Reply:
Changing this to an async function causes the web platform tests to fail in rather fun ways... 

![image](https://user-images.githubusercontent.com/439929/123158810-1702c980-d421-11eb-8da0-89ae7880e4c0.png)

I presume it is because of the additional Promises and microtasks that are created...